File: pkg/remote/mdcs_remote/models/control.py
import enum

from sqlalchemy import Column, ForeignKey, String, Enum
from sqlalchemy.orm import relationship

from .generic import Model


class ControlType(enum.Enum):
    NONE = enum.auto()
    BUTTON = enum.auto()
    COLOR = enum.auto()


class Control(Model):
    """
    A parent table for control set controls. Controls are a complete disjoint specialization.
    """

    __tablename__ = 'control'

    uuid = Column(String(22), primary_key=True)
    controlset_uuid = Column(String(22), ForeignKey('controlset.uuid'), nullable=False)
    type = Column(Enum(ControlType), nullable=False)
    description = Column(String(64), nullable=False, default="")

    controlset = relationship('ControlSet', back_populates='controls')

    __mapper_args__ = {
        'polymorphic_identity': ControlType.NONE,
        'polymorphic_on': type
    }

    def __repr__(self):
        return "<Control(uuid='{0}', controlset_uuid='{1}', type=ControlType.{2}, description='{3}')>".format(
            self.uuid,
            self.controlset_uuid,
            self.type.name,
            self.description)

File: pkg/remote-mobile/lib/models/control.dart
import 'package:remote/models/model.dart';

/// Control type.
enum ControlType {
  none,
  button,
  color
}

/// A generic control.
class Control extends Model {
  final String uuid;
  final String controlSetUuid;
  final String name;
  final ControlType type;
  final String description;

  Control({this.uuid, this.controlSetUuid, this.name, this.type, this.description});
  Control.fromJSON(Map<String, dynamic> data, {String controlSetUuid}):
    this.uuid = data['uuid'] as String,
    this.controlSetUuid = controlSetUuid != null ? controlSetUuid : data['controlset_uuid'] as String,
    this.name = data['name'] as String,
    this.type = Control.parseControlType(data['type'] as String),
    this.description = data['description'] as String;

  static ControlType parseControlType(String value) {
    switch (value.toLowerCase()) {
      case 'none': return ControlType.none;
      case 'button': return ControlType.button;
      case 'color': return ControlType.color;
      default: throw Exception('unknown control type: $value');
    }
  }

  @override
  String get primaryKey => this.uuid;
}

/// A generic control value.
class ControlValue {
  /// Get the JSON serializable representation of the value.
  Map<String, dynamic> toData() => {};
}
